### PR TITLE
Add multiple bq column functionality, custom user id column, and multi event tables

### PR DESCRIPTION
**Pull request by rlh1994**
## Description & motivation
Adds 3 new features across 3 commits (will rebase, not squash).

1. Add the option for a custom `user_id` field that can come from the a sde or a context. The macro is updated to support new columns, the python script manages the reading of these and a new config schema version ensures it matches. Tests updated in both pytest and macros.

2. Add mutli-event functionality to the tables, either default or mutliple sdes. Adds an alias field for sdes and converts the arguments in the configs to lists - causing a major change to the config schema version number. Tests are updated and new tests added to cover these.

3. Due to the issue of BQ schema versioning the original script only used the specific version of the schema that was provided in the config. This change makes it so that it now uses all columns under the same major version, aligning the functionality with the rest of the dbt packages we provide. Because the macro that combines the columns has to query the database to do this, I have had to seed the bare minimum data required to generate a valid `_this_run` table for the macro to hit. This still does nothing to test the incremental logic, and the data is not meant to be realistic, instead it is there to test the generated sql remains correct.

I would suggest reviewing each commit individually as it will make it easier to follow, although things to get reworked in later commits.

## Checklist
- [x] I have verified that these changes work locally
- [x] I have updated the README.md (if applicable)
- [x] I have added tests & descriptions to my models (and macros if applicable)
- [x] I have raised a [documentation](https://github.com/snowplow/documentation) PR if applicable (Will do as part of release)

### CODE REVIEW

**Review by agnessnowplow — COMMENTED** [review]
I started reviewing the custom userid field support one first. Apart from a handful of negligible wording change recommendations I have found one issue with running this in Snowflake. It leaves the camel case for the standard user_id field, which is snake_case in snowflake, too so it will fail building the model:
image

**Comment by agnessnowplow** [comment]
I will continue the review tomorrow but one more thing I have noticed (although it is not related to this commit) is when  I run the unit tests locally (and also checking the[ Pytest details ](https://github.com/snowplow/dbt-snowplow-normalize/actions/runs/3657779336/jobs/6181754836#step:5:71)on GitHub) it fails the `test_validate_json`. Thought I'd mention already so that you can take a look.

**Comment by rlh1994** [comment]
>It leaves the camel case for the standard user_id field, which is snake_case in snowflake, too so it will fail building the model:
I think this is actually the case across all warehouses - I had assumed if you were specifying a default column that you would put it in the format of the table, but given all the standard columns (as far as I know) are snakecase across all warehouses it would make sense to add that convertion regardless, thanks!

File: macros/schema.yml [hunk 1]
```diff
@@ -37,6 +37,15 @@ macros:
   - name: users_table
     description: A macro to produce a users table from the `base_events_this_run` table, using the latest context values as defined by the collector_tstamp.
     arguments:
+      - name: user_id_field
```
[the remark this answers is not in the document]
Comment (rlh1994) [reply macros/schema.yml:40]: Agree, I'll add a warning raise in both the python and the macro so at least they get notice of it should it happen.

**Comment by georgewoodhead** [comment]
I've reviewed feature 2 (Add mutli-event functionality to the tables) and all looks good! If being ⛏️   could we handle or add a  warning when alias length not equal to sde/context length, e.g. if you set 1 context alias but had 2 contexts in the config, the 2nd context would have columns starting with _? This package is awesome btw 😎

**Comment by rlh1994** [comment]
>could we handle or add a warning when alias length not equal to sde/context length, e.g. if you set 1 context alias but had 2 contexts in the config, the 2nd context would have columns starting with _?

Good idea, I've stuck it in as a raised warning in this version (as there is no way to do this via jsonschema) but will add a ticket in JIRA to add support for empty string inputs here in a future version (as it will require more edits to the macros and new tests etc).

**Comment by georgewoodhead** [comment]
If someone uses the same event name in more than one event like 
```
 "events":[
        {
            "event_names": ["page_view", "link_click"],
            ...
        },
        {
            "event_names": ["link_click"],
            ...
        }
```
the model will run first time, but will fail on any incremental runs when it tries to update/merge `snowplow_events_normalized` because we will have duplicate event_ids:
image

I can't immediately think of a use case when a user might want to have a setup like this so I don't think we need to have a work around for this, but it could be accidentally done and not immediately obvious why their model failed. Could we add a exception in the python script to say an event name can only be used once?

**Comment by rlh1994** [comment]
> I can't immediately think of a use case when a user might want to have a setup like this so I don't think we need to have a work around for this, but it could be accidentally done and not immediately obvious why their model failed. Could we add a exception in the python script to say an event name can only be used once?

Good catch! There was already a check for duplicate table names but not for something like this. I've added a concatinated field (I haven't used the surrogate key generator as neither column should ever be null and this makes it a lot easier to give upgrade instructions to people as this is a breaking change if they don't want to re-run everything) on the off chance anyone does every want to do this.

**Comment by georgewoodhead** [comment]
Feature 3 (multiple bigquery column functionality) LGTM!

**Review by georgewoodhead — APPROVED** [review]

**Pull request merged**